### PR TITLE
Make user tokens types configurable

### PR DESCRIPTION
**Pull request by batopa**
This PR allow to add user tokens types with configuration.

### CODE REVIEW

File: plugins/BEdita/Core/src/Model/Table/UserTokensTable.php [hunk 1]
```diff
@@ -38,11 +39,11 @@
 class UserTokensTable extends Table
 {
     /**
-     * List of allowed token types
+     * List of default allowed token types
      *
      * @var array
      */
-    const TOKEN_TPYES = ['otp', 'refresh', 'recovery', '2fa', 'access'];
+    const DEFAULT_TOKEN_TPYES = ['otp', 'refresh', 'recovery', '2fa', 'access'];
```
Comment (fquffio) [inline plugins/BEdita/Core/src/Model/Table/UserTokensTable.php]: Typo: `TPYES`

File: plugins/BEdita/Core/src/Model/Table/UserTokensTable.php [hunk 2]
```diff
@@ -127,6 +128,18 @@ public function buildRules(RulesChecker $rules)
...
     }
 
+    /**
+     * Return the list of allowed token types merging default with configured.
+     *
+     * @return array
+     */
+    public function getTokenTypes()
+    {
+        $confTypes = (array)Configure::read('UserTokens');
```
Comment (fquffio) [inline plugins/BEdita/Core/src/Model/Table/UserTokensTable.php]: The `UserTokens` configuration section might include more keys in the future. I'd use `UserTokens.types` instead.

File: plugins/BEdita/Core/tests/TestCase/Model/Table/UserTokensTableTest.php [hunk 3]
```diff
@@ -114,4 +115,24 @@ public function testValidFinder()
...
+        $expected = UserTokensTable::DEFAULT_TOKEN_TPYES;
+        static::assertEquals($expected, $this->UserTokens->getTokenTypes());
+
+        $conf = ['token_one', 'token_two'];
+        Configure::write('UserTokens', $conf);
+        $expected = array_merge($expected, $conf);
+        static::assertEquals($expected, $this->UserTokens->getTokenTypes());
+
+        Configure::write('UserTokens', ['token_one', 'token_two', 'access']);
+        static::assertEquals($expected, $this->UserTokens->getTokenTypes());
```
Comment (fquffio) [inline plugins/BEdita/Core/tests/TestCase/Model/Table/UserTokensTableTest.php:173]: I know it's a bit overkill in this case, but if possible I'd refactor this test to use data provider.

**Review by fquffio — CHANGES_REQUESTED** [review]

File: plugins/BEdita/Core/src/Model/Table/UserTokensTable.php (hunk 1 above)
Comment (batopa) [reply plugins/BEdita/Core/src/Model/Table/UserTokensTable.php]: ups.. right :)

File: plugins/BEdita/Core/src/Model/Table/UserTokensTable.php (hunk 2 above)
Comment (batopa) [reply plugins/BEdita/Core/src/Model/Table/UserTokensTable.php]: yeah, maybe it's better

**Review by stefanorosanelli — APPROVED** [review]

File: plugins/BEdita/Core/tests/TestCase/Model/Table/UserTokensTableTest.php (hunk 3 above)
Comment (batopa) [reply plugins/BEdita/Core/tests/TestCase/Model/Table/UserTokensTableTest.php:173]: ahahah.... I had already refactored removing previous provider to simplify... re-re-factor is on the way :+1:

Comment (fquffio) [reply plugins/BEdita/Core/tests/TestCase/Model/Table/UserTokensTableTest.php:173]: 😄 Sorry!!

**Review by fquffio — APPROVED** [review]

**Pull request merged**